pkg/controller/endpointslice/reconciler.go

```diff
@@ -0,0 +1,297 @@
+			}
+		}
+	}
+
+	// When no endpoint slices would usually exist, we need to add a placeholder.
+	if len(existingSlices) == sliceNamesToDelete.Len() && len(slicesToCreate) < 1 {
+		placeholderSlice := newEndpointSlice(service, &endpointMeta{Ports: []discovery.EndpointPort{}})
+		slicesToCreate = append(slicesToCreate, placeholderSlice)
+	}
+
+	return r.finalize(service, slicesToCreate, slicesToUpdate, sliceNamesToDelete, triggerTime)
+}
+
+// finalize creates, updates, and deletes slices as specified
+func (r *reconciler) finalize(
+	service *corev1.Service,
+	slicesToCreate,
+	slicesToUpdate []*discovery.EndpointSlice,
+	sliceNamesToDelete sets.String,
+	triggerTime time.Time,
+) error {
+	errs := []error{}
+
+	// If there are slices to create and delete, change the creates to updates
+	// of the slices that would otherwise be deleted.
+	for len(slicesToCreate) > 0 && sliceNamesToDelete.Len() > 0 {
+		sliceName, _ := sliceNamesToDelete.PopAny()
+		slice := slicesToCreate[len(slicesToCreate)-1]
+		slicesToCreate = slicesToCreate[:len(slicesToCreate)-1]
+		slice.Name = sliceName
+		slicesToUpdate = append(slicesToUpdate, slice)
+	}
+
+	for _, endpointSlice := range slicesToCreate {
+		addTriggerTimeAnnotation(endpointSlice, triggerTime)
+		_, err := r.client.DiscoveryV1alpha1().EndpointSlices(service.Namespace).Create(endpointSlice)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("Error creating EndpointSlice for Service %s/%s: %v", service.Namespace, service.Name, err))
+		}
+	}
+
+	for _, endpointSlice := range slicesToUpdate {
+		addTriggerTimeAnnotation(endpointSlice, triggerTime)
+		_, err := r.client.DiscoveryV1alpha1().EndpointSlices(service.Namespace).Update(endpointSlice)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("Error updating %s EndpointSlice for Service %s/%s: %v", endpointSlice.Name, service.Namespace, service.Name, err))
+		}
+	}
+
+	for sliceNamesToDelete.Len() > 0 {
+		sliceName, _ := sliceNamesToDelete.PopAny()
+		err := r.client.DiscoveryV1alpha1().EndpointSlices(service.Namespace).Delete(sliceName, &metav1.DeleteOptions{})
+		if err != nil {
+			errs = append(errs, fmt.Errorf("Error deleting %s EndpointSlice for Service %s/%s: %v", sliceName, service.Namespace, service.Name, err))
+		}
+	}
+
+	return utilerrors.NewAggregate(errs)
+}
+
+// reconcileByPortMapping compares the endpoints found in existing slices with
+// the list of desired endpoints and returns lists of slices to create, update,
+// and delete. The logic is split up into several main steps:
+// 1. Iterate through existing slices, delete endpoints that are no longer
+//    desired and update matching endpoints that have changed.
+// 2. Iterate through slices that have been modified in 1 and fill them up with
+//    any remaining desired endpoints.
+// 3. If there still desired endpoints left, try to fit them into a previously
+//    unchanged slice and/or create new ones.
+func (r *reconciler) reconcileByPortMapping(
```

Comment:
The below will work. 

My concern is `fragmentation` and `lack of compaction`. Do we want to favor `large # of slices with small # of endpoints` or `small # of slices with large # of endpoints`? in a scenario where the # of pods fronted by the same services grows multiple orders of magnitude then back down the below will leave behind many endpoint slices with very little utilization each.